JOING-47 feat: 회원 관리 API 구현

hyoom opened the pull request:
## #️⃣ 연관된 이슈

> #JOING-47

<br/>

## 📝 작업 내용

> 마이페이지에서 사용될 회원 정보 조회 기능과 회원 정보 수정 기능을 구현하였습니다.

<br/>

## 📷 스크린샷(선택)

> 이미지가 있다면 첨부해주세요

<br/>

## 💬 리뷰 요구사항(선택)

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
>
> ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요?
>

hyoom committed:
JOING-47 feat: 회원 정보 조회 기능 구현

hyoom committed:
JOING-47 feat: 회원 정보 수정 기능 구현

hyoom committed:
Merge branch 'develop' into feature/JOING-47-user-info-api

jayden-bin commented on src/main/java/com/ktb/joing/domain/user/controller/UserController.java:
```diff
@@ -38,4 +42,18 @@ public ResponseEntity<Void> productManagerSignUp(
 
+    @GetMapping
+    public ResponseEntity<UserResponse<?>> getUserInfo(@AuthenticationPrincipal CustomOAuth2User customOAuth2User) {
+        UserResponse<?> response = userService.getUser(customOAuth2User.getUsername());
```
\<?> 는 뭐가 반환될지 예상할수 없어서 지양하시는게 좋습니다..!
제네릭은 되도록이면 명시적으로 적어주시면 좋을것 같습니다~!

jayden-bin commented on src/main/java/com/ktb/joing/domain/user/service/UserService.java:
```diff
@@ -77,6 +83,41 @@ public void productManagerSignUp(String username, ProductManagerSignupRequest re
     }
 
+    // 회원 정보 조회
+    @Transactional(readOnly = true)
```
메소드 내부에 쿼리가 단 하나뿐이고 실행이 오래걸리는 코드(API호출)같은게 없다면 트랜잭션을 거는 메리트가 거의 없어요!

jayden-bin commented on src/main/java/com/ktb/joing/domain/user/service/UserService.java:
```diff
@@ -77,6 +83,41 @@ public void productManagerSignUp(String username, ProductManagerSignupRequest re
+    }
+
+    // 회원 정보 수정
+    public UserResponse<?> updateUser(String username, UserUpdateRequest request) {
+        User user = userRepository.findByUsername(username)
+                .orElseThrow(() -> new UserException(UserErrorCode.USER_NOT_FOUND));
+
+        if (request.getNickname() != null) {
+            validateDuplicateNickname(request.getNickname());
+        }
+
+        if (user instanceof Creator creator) {
+            creator.update(request);
+            return new UserResponse<>(UserType.CREATOR,
+                    CreatorResponse.builder().creator(creator).build());
+        }
+
+        ProductManager productManager = (ProductManager) user;
+        productManager.update(request);
+        return new UserResponse<>(UserType.PRODUCT_MANAGER,
+                ProductManagerResponse.builder().productManager(productManager).build());
+    }
```
요 메소드에서 명시적으로 save하는 코드도 없고 JPA 더티체킹을 쓰는것 같지도 않아보이는데
변경사항이 제대로 DB에 반영되나요??

jayden-bin replied:
앗 ㅋㅋㅋ 요건 제가 클래스에 붙어있는 Transactional을 못봤네요
그런데 메소드 단위로 트랜잭션을 선언해주시는게 이런 오해를 일으키지 않고 안전합니다~!

hyoom committed:
Merge branch 'develop' into feature/JOING-47-user-info-api

jayden-bin reviewed (commented):
<img width="641" alt="스크린샷 2024-12-03 오후 3 59 51" src="https://github.com/user-attachments/assets/de153a3f-16fc-422b-976c-776d11312f0c">

.DS_STORE 파일들은 깃헙에 올라오지 않도록 ignore 처리해두는게 좋습니다~!

hyoom committed:
JOING-47 refactor: UserService 메서드별 트랜잭션 분리

- 클래스 레벨의 @Transactional 제거
- 각 메서드 특성에 맞게 @Transactional 개별 적용
- 단일 쿼리만 수행하는 조회 메서드(getUser)에서는 트랜잭션이 불필요하므로 제거

hyoom committed:
JOING-47 chore: .gitignore에 .DS_Store 추가

hyoom committed:
JOING-47 feat: OAuth2 로그인 응답에 사용자 타입 추가

- CustomSuccessHandler에서 리다이렉트 시 type 파라미터 추가
- DiscriminatorValue를 "CREATOR", "PRODUCT_MANAGER"로 수정

hyoom committed:
JOING-47 feat: 회원가입 응답에 SignupResponse 추가

- 회원가입 API 응답에 userType(크리에이터/기획자) 정보 포함

hyoom committed:
JOING-47 refactor: 제네릭 와일드카드를 명시적 타입으로 변경

- UserResponse<?> 제거
- 타입별 엔드포인트 분리(/creator, /productmanager)

jayden-bin commented on src/main/java/com/ktb/joing/domain/user/dto/request/UserUpdateRequest.java:
```diff
@@ -0,0 +1,38 @@
+    @Size(max = 20, message = "닉네임은 최대 20자까지 가능합니다")
+    private String nickname;
+
+    @Email(regexp = "^[A-Za-z0-9+_.-]+@(.+)$", message = "올바른 이메일 형식이 아닙니다")
```
\@Email만써도 동작할텐데 regexp를 사용해준 이유가 있을까요??
그리고 regexp를 쓸거라면 @ Pattern을 쓰는게 더 직관적입니다~!

hyoom replied:
말씀해 주신대로 @Email 어노테이션은 메일 검증을 내장하고 있어서 추가적인 정규식이 불필요했네요! 몰랐습니다... @Email만 사용하도록 수정하겠습니다. 좋은 피드백 감사합니다!

hyoom committed:
JOING-47 refactor: @Email 어노테이션의 불필요한 정규식 제거

hyoom committed:
JOING-47 refactor: Update DTO를 유저 타입별로 분리

- UserUpdateRequest 제거
- CreatorUpdateRequest, ProductManagerUpdateRequest로 분리

hyoom committed:
JOING-47 feat: 크리에이터 프로필 평가 API 파라미터 수정
